src/mrpro/operators/_SliceProjectionOp.py

```diff
@@ -0,0 +1,449 @@
+            case (None, None):
+                raise RuntimeError('Either matrix or matrix adjoint must be set')
+            case (matrix, None) if matrix is not None:
+                matrix_adjoint = matrix.H
+            case (None, matrix_adjoint) if matrix_adjoint is not None:
+                matrix = matrix_adjoint.H
+            case (matrix, matrix_adjoint):
+                ...
+
+        # For the (unusual case) of batched volumes, we will apply for each element in series
+        xflat = torch.atleast_2d(einops.rearrange(x, '... x y z -> (...) (x y z)'))
+        y = torch.stack(
+            [_MatrixMultiplication.apply(x, matrix, matrix_adjoint).reshape(self._range_shape) for x in xflat], -4
+        )
+        y = y.reshape(*y.shape[:-4], *x.shape[:-3], *y.shape[-3:])
+        return (y,)
+
+    def adjoint(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Transform from a 2D slice to a 3D Volume."""
+        match (self.matrix, self.matrix_adjoint):
+            # selection based on the optimize_for setting
+            case (None, None):
+                raise RuntimeError('Either matrix or matrix adjoint must be set')
+            case (matrix, None) if matrix is not None:
+                matrix_adjoint = matrix.H
+            case (None, matrix_adjoint) if matrix_adjoint is not None:
+                matrix = matrix_adjoint.H
+            case (matrix, matrix_adjoint):
+                ...
+
+        # For the (unusual case) of batched volumes, we will apply for each element in series
+        n_batchdim = len(self._range_shape) - 3
+        # x_domainbatch_range has all volume batch dimensions moved to the front
+        x_domainbatch_range = x.moveaxis(tuple(range(n_batchdim, x.ndim - 3)), tuple(range(x.ndim - 3 - n_batchdim)))
+        # x_flatdomainbatch_flatrange is 2D with shape
+        # (all batch dimensions of the volume flattened, all range dimensions flattened)
+        x_domainbatch_flatrange = torch.atleast_2d(x_domainbatch_range.flatten(start_dim=-len(self._range_shape)))
+        x_flatdomainbatch_flatrange = x_domainbatch_flatrange.flatten(end_dim=-2)
+        # y_flatdomainbatch has shape (all batch dimensions of the volume flattened, *range dimensions)
+        y_flatdomainbatch = torch.stack(
+            [
+                _MatrixMultiplication.apply(x, matrix_adjoint, matrix).reshape(self._domain_shape)
+                for x in x_flatdomainbatch_flatrange
+            ]
+        )
+        y = y_flatdomainbatch.reshape(*x.shape[n_batchdim:-3], *y_flatdomainbatch.shape[1:])
+        return (y,)
+
+    @staticmethod
+    def join_matrices(matrices: Sequence[torch.Tensor]) -> torch.Tensor:
+        values = []
+        target = []
+        source = []
+        for i, m in enumerate(matrices):
+            if not m.shape == matrices[0].shape:
+                raise ValueError('all matrices should have the same shape')
+            c = m.coalesce()
```

Comment:
Could you add some comments on what coalesce() does here? Maybe superfluous once there is a docstring.